Fix before sunrise OR after sunset condition

janick opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

The following condition does not work:

```
condition: sun
    before: sunrise
    after: sunset
```

because the code does an AND operation from midnight to midnight, and it is not possible for the condition to be TRUE both before sunrise and after sunset. What is needed is to catch this condition and treat is as an OR. That is what this PR does.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #70781 
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/24005

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [ ] The code change is tested and works locally.
- [ ] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [ ] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

janick committed:
Fix Issue #70781

janick committed:
Remove spaces on blank line

janick committed:
Added tests for before sunrise or after sunset

I have to way to check syntax, so crossing fingers...

janick committed:
Fix blank lines

veleek commented on homeassistant/helpers/condition.py:
```diff
@@ -615,6 +615,14 @@ def sun(
         return False
 
+    # Special case: before sunrise OR after sunset
+    if before == SUN_EVENT_SUNRISE and after == SUN_EVENT_SUNSET:
```
In the case where there is no sunrise or sunset, I would expect this OR condition to still fire, so this should probably go before the two `if blah is Non` statements above.

janick replied:
Yes, you are correct.

veleek reviewed (changes requested):
Move condition as per comments.

veleek committed:
Improve understandability of sun condition logic

Move a few things around, add/remove some temporary variables with the goal of making the (complicated) sunrise/sunset logic in the sun condition a little bit more understandable.

veleek commented:
@janick - I was trying to review this and make sure that I completely understood all the edge cases here and was having a hard time making sure that I had the logic right in my head.  

I made a quick pass at reworking some of the initial code in the condition to help make it a bit more obvious what's actually going on ([currently in a branch on my fork](https://github.com/home-assistant/core/compare/dev...veleek:home-assistant:benran/sun-condition#diff-0558c1da8b70e0262af39161c04b536bb9398f7963b41f9338f9e85cee012f42)).  This would be in addition to your change (or potentially added afterwards).  

Take a look and see if you think it's worth adding, if so I can PR directly into your branch.

janick committed:
Move special case condition up

Handles the case where the sun sets/rises but does not rise/set in the same day.

janick commented:
@veleek  Please do!

veleek commented:
Done! https://github.com/janick/core/pull/1

veleek commented on homeassistant/helpers/condition.py:
```diff
@@ -586,31 +586,46 @@ def sun(
+    has_sunrise_condition = SUN_EVENT_SUNRISE in (before, after)
+    has_sunset_condition = SUN_EVENT_SUNSET in (before, after)
+
+    after_sunrise = today > dt_util.as_local(cast(datetime, sunrise)).date()
+    if after_sunrise and has_sunrise_condition:
+        tomorrow = today + timedelta(days=1)
+        sunrise = get_astral_event_date(hass, SUN_EVENT_SUNRISE, tomorrow)
+
+    after_sunset = today > dt_util.as_local(cast(datetime, sunset)).date()
+    if after_sunset and has_sunset_condition:
+        tomorrow = today + timedelta(days=1)
+        sunset = get_astral_event_date(hass, SUN_EVENT_SUNSET, tomorrow)
```
After an exhaustive set of tests to figure out what these were for, I came to the conclusion that as far as I can tell these blocks will never get hit...

`today` is the date for the local time.  As far as I can tell (looking at the [astral source code this value comes from](https://github.com/sffjunkie/astral/blob/master/src/astral/sun.py#L761-L762)), there's no world in which the sunrise or sunset for a given `today` will ever be return for tomorrow.

I think you can just delete these.  At least my tests still all pass, but I haven't does exhaustive tests which include offsets yet, but since `today` and `sunrise`/`sunset` aren't influenced by the offset yet, that shouldn't have any impact.

frenck replied:
I'm not sure, but let's keep them out of scope for this particular issue.

veleek commented on tests/helpers/test_condition.py:
```diff
@@ -2837,6 +2837,128 @@ async def test_if_action_before_and_after_during(hass, hass_ws_client, calls):
     )
 
 
+async def test_if_action_before_or_after_during(hass, hass_ws_client, calls):
```
To test this locally myself, I made a [parameterized test](https://docs.pytest.org/en/6.2.x/parametrize.html) so I could easily test out a bunch of variations with a single test, and I also just changed it to call `condition.sun` directly instead of creating the automation so the tests run faster.  

With a parameterized test, I suppose it doesn't matter whether the test is slightly longer with the automation components, but it could still help to simplify the test duplication.

```python
@pytest.mark.parametrize(
    "now_str, after, before, expected",
    [
        ( "2015-09-16 07:00:00+0000 UTC", SUN_EVENT_SUNRISE, None, False ), # After Sunrise @ Midnight
        ( "2015-09-16 12:33:18+0000 UTC", SUN_EVENT_SUNRISE, None, False ), # After Sunrise @ Sunrise - 1h
        ( "2015-09-16 13:32:18+0000 UTC", SUN_EVENT_SUNRISE, None, False ), # After Sunrise @ Sunrise - 1m
        ( "2015-09-16 13:33:17+0000 UTC", SUN_EVENT_SUNRISE, None, False ), # After Sunrise @ Sunrise - 1s
        ( "2015-09-16 13:33:18+0000 UTC", SUN_EVENT_SUNRISE, None, False ), # After Sunrise @ Sunrise
        ( "2015-09-16 13:33:19+0000 UTC", SUN_EVENT_SUNRISE, None, True ),  # After Sunrise @ Sunrise + 1s
        ( "2015-09-16 13:34:18+0000 UTC", SUN_EVENT_SUNRISE, None, True ),  # After Sunrise @ Sunrise + 1m
        ( "2015-09-16 14:33:18+0000 UTC", SUN_EVENT_SUNRISE, None, True ),  # After Sunrise @ Sunrise + 1h
        ( "2015-09-17 00:53:45+0000 UTC", SUN_EVENT_SUNRISE, None, True ),  # After Sunrise @ Sunset - 1h
        ( "2015-09-17 01:52:45+0000 UTC", SUN_EVENT_SUNRISE, None, True ),  # After Sunrise @ Sunset - 1m
        ( "2015-09-17 01:53:44+0000 UTC", SUN_EVENT_SUNRISE, None, True ),  # After Sunrise @ Sunset - 1s
        ( "2015-09-17 01:53:45+0000 UTC", SUN_EVENT_SUNRISE, None, True ),  # After Sunrise @ Sunset
        ( "2015-09-17 01:53:46+0000 UTC", SUN_EVENT_SUNRISE, None, True ),  # After Sunrise @ Sunset + 1s
        ( "2015-09-17 01:54:45+0000 UTC", SUN_EVENT_SUNRISE, None, True ),  # After Sunrise @ Sunset + 1m
        ( "2015-09-17 02:53:45+0000 UTC", SUN_EVENT_SUNRISE, None, True ),  # After Sunrise @ Sunset + 1h
        ( "2015-09-17 06:59:59+0000 UTC", SUN_EVENT_SUNRISE, None, True ),  # After Sunrise @ Midnight - 1s next day

        ( "2015-09-16 07:00:00+0000 UTC", SUN_EVENT_SUNSET, None, False ),  # After Sunset @ Midnight
        ( "2015-09-16 12:33:18+0000 UTC", SUN_EVENT_SUNSET, None, False ),  # After Sunset @ Sunrise - 1h
        ( "2015-09-16 13:32:18+0000 UTC", SUN_EVENT_SUNSET, None, False ),  # After Sunset @ Sunrise - 1m
        ( "2015-09-16 13:33:17+0000 UTC", SUN_EVENT_SUNSET, None, False ),  # After Sunset @ Sunrise - 1s
        ( "2015-09-16 13:33:18+0000 UTC", SUN_EVENT_SUNSET, None, False ),  # After Sunset @ Sunrise
        ( "2015-09-16 13:33:19+0000 UTC", SUN_EVENT_SUNSET, None, False ),  # After Sunset @ Sunrise + 1s
        ( "2015-09-16 13:34:18+0000 UTC", SUN_EVENT_SUNSET, None, False ),  # After Sunset @ Sunrise + 1m
        ( "2015-09-16 14:33:18+0000 UTC", SUN_EVENT_SUNSET, None, False ),  # After Sunset @ Sunrise + 1h
        ( "2015-09-17 00:53:45+0000 UTC", SUN_EVENT_SUNSET, None, False ),  # After Sunset @ Sunset - 1h
        ( "2015-09-17 01:52:45+0000 UTC", SUN_EVENT_SUNSET, None, False ),  # After Sunset @ Sunset - 1m
        ( "2015-09-17 01:53:44+0000 UTC", SUN_EVENT_SUNSET, None, False ),  # After Sunset @ Sunset - 1s
        ( "2015-09-17 01:53:45+0000 UTC", SUN_EVENT_SUNSET, None, True ),   # After Sunset @ Sunset
        ( "2015-09-17 01:53:46+0000 UTC", SUN_EVENT_SUNSET, None, True ),   # After Sunset @ Sunset + 1s
        ( "2015-09-17 01:54:45+0000 UTC", SUN_EVENT_SUNSET, None, True ),   # After Sunset @ Sunset + 1m
        ( "2015-09-17 02:53:45+0000 UTC", SUN_EVENT_SUNSET, None, True ),   # After Sunset @ Sunset + 1h
        ( "2015-09-17 06:59:59+0000 UTC", SUN_EVENT_SUNSET, None, True ),   # After Sunset @ Midnight - 1s next day

        ( "2015-09-16 07:00:00+0000 UTC", SUN_EVENT_SUNRISE, SUN_EVENT_SUNSET, False ), # Daylight  @ Midnight
        ( "2015-09-16 12:33:18+0000 UTC", SUN_EVENT_SUNRISE, SUN_EVENT_SUNSET, False ), # Daylight  @ Sunrise - 1h
        ( "2015-09-16 13:32:18+0000 UTC", SUN_EVENT_SUNRISE, SUN_EVENT_SUNSET, False ), # Daylight  @ Sunrise - 1m
        ( "2015-09-16 13:33:17+0000 UTC", SUN_EVENT_SUNRISE, SUN_EVENT_SUNSET, False ), # Daylight  @ Sunrise - 1s
        ( "2015-09-16 13:33:18+0000 UTC", SUN_EVENT_SUNRISE, SUN_EVENT_SUNSET, False ), # Daylight  @ Sunrise
        ( "2015-09-16 13:33:19+0000 UTC", SUN_EVENT_SUNRISE, SUN_EVENT_SUNSET, True ),  # Daylight  @ Sunrise + 1s
        ( "2015-09-16 13:34:18+0000 UTC", SUN_EVENT_SUNRISE, SUN_EVENT_SUNSET, True ),  # Daylight  @ Sunrise + 1m
        ( "2015-09-16 14:33:18+0000 UTC", SUN_EVENT_SUNRISE, SUN_EVENT_SUNSET, True ),  # Daylight  @ Sunrise + 1h
        ( "2015-09-17 00:53:45+0000 UTC", SUN_EVENT_SUNRISE, SUN_EVENT_SUNSET, True ),  # Daylight  @ Sunset - 1h
        ( "2015-09-17 01:52:45+0000 UTC", SUN_EVENT_SUNRISE, SUN_EVENT_SUNSET, True ),  # Daylight  @ Sunset - 1m
        ( "2015-09-17 01:53:44+0000 UTC", SUN_EVENT_SUNRISE, SUN_EVENT_SUNSET, True ),  # Daylight  @ Sunset - 1s
        ( "2015-09-17 01:53:45+0000 UTC", SUN_EVENT_SUNRISE, SUN_EVENT_SUNSET, False ), # Daylight  @ Sunset
        ( "2015-09-17 01:53:46+0000 UTC", SUN_EVENT_SUNRISE, SUN_EVENT_SUNSET, False ), # Daylight  @ Sunset + 1s
        ( "2015-09-17 01:54:45+0000 UTC", SUN_EVENT_SUNRISE, SUN_EVENT_SUNSET, False ), # Daylight  @ Sunset + 1m
        ( "2015-09-17 02:53:45+0000 UTC", SUN_EVENT_SUNRISE, SUN_EVENT_SUNSET, False ), # Daylight  @ Sunset + 1h
        ( "2015-09-17 06:59:59+0000 UTC", SUN_EVENT_SUNRISE, SUN_EVENT_SUNSET, False ), # Daylight  @ Midnight - 1s next day
    ],
)
async def test_condition_blahsun(hass, now_str, after, before, expected):
    """
    A test that verifies the behavior of the sun condition

    # sunrise: 2015-09-16 06:33:18 local, sunset: 2015-09-16 18:53:45 local
    # sunrise: 2015-09-16 13:33:18 UTC,   sunset: 2015-09-17 01:53:45 UTC
    """

    time_format = "%Y-%m-%d %H:%M:%S%z %Z"
    now = datetime.strptime(now_str, time_format)
    print(now.tzinfo)

    with patch("homeassistant.util.dt.utcnow", return_value=now):
        result = condition.sun(hass, before, after, None, None)
        assert result == expected

        # When we swap the before and after conditions, the result should be opposite
        result = condition.sun(hass, after, before, None, None)
        assert result != expected
```

Feel free to steal this if you think it's useful.  Could probably also get rid of a bunch of the extra cases (like -1h, -1m, +1m, +1h), I was just trying to be thorough but I don't think they're really necessary.

janick replied:
@veleek  Did you test the "nighttime" condition as well? I don't see any call with "before sunrise, after sunset" but it may be simply an excerpt issue.

I'll gladly merge in your parameterized tests in my PR.

veleek replied:
That's actually covered by the "swap" portion of the test.  As I was duplicating test cases, I realized that it was essentially identical to just swapping before and after and swapping the expected result.  I couldn't think of a good way to name stuff or indicate that otherwise.  Could just duplicate the test cases if we wanted.

```python
    with patch("homeassistant.util.dt.utcnow", return_value=now):
        result = condition. Sun(hass, before, after, None, None)
        assert result == expected

        # When we swap the before and after conditions, the result should be opposite
        result = condition. Sun(hass, after, before, None, None)
        assert result != expected
```

janick replied:
@veleek Can you submit a PR with your testcase changes?

veleek replied:
I'll try to do that in the next day or two if I have some time!

emontnemery commented:
There have been many previous attempts to get the sun conditions right, so we need to very clear about what needs to be changed and why it's (hopefully) correct this time.

Hence, please update the PR description to explain very clearly what has changed and why.
Ideally, give examples of what wasn't working before but is working now with these changes.

janick commented:
@emontnemery Done

janick committed:
Removed trailing space

frenck reviewed (approved):
This is very nice @janick 👍 🥇 

The test improvements/simplification are welcome (if you are up for another PR). But going to merge this one to get it into the upcoming beta.

frenck commented:
Rebased on the latest dev to fix an CI error.

emontnemery reviewed (approved):
Thanks a lot @janick, I agree with @frenck this is very nice 🥇

Please open follow-up PRs with the suggested test improvements and simplification of code 👍

frenck reviewed (changes requested):
@janick The tests are failing, could you take a look?

emontnemery commented on tests/helpers/test_condition.py:
```diff
@@ -2837,6 +2837,128 @@ async def test_if_action_before_and_after_during(hass, hass_ws_client, calls):
+    with patch("homeassistant.util.dt.utcnow", return_value=now):
+        hass.bus.async_fire("test_event")
+        await hass.async_block_till_done()
+        assert len(calls) == 0
```
```suggestion
        assert len(calls) == 1
```

janick replied:
Look good

emontnemery commented on tests/helpers/test_condition.py:
```diff
@@ -2837,6 +2837,128 @@ async def test_if_action_before_and_after_during(hass, hass_ws_client, calls):
+    with patch("homeassistant.util.dt.utcnow", return_value=now):
+        hass.bus.async_fire("test_event")
+        await hass.async_block_till_done()
+        assert len(calls) == 0
```
```suggestion
        assert len(calls) == 2
```

janick replied:
Look good

emontnemery commented on tests/helpers/test_condition.py:
```diff
@@ -2837,6 +2837,128 @@ async def test_if_action_before_and_after_during(hass, hass_ws_client, calls):
+    with patch("homeassistant.util.dt.utcnow", return_value=now):
+        hass.bus.async_fire("test_event")
+        await hass.async_block_till_done()
+        assert len(calls) == 1
```
```suggestion
        assert len(calls) == 2
```

janick replied:
Look good

emontnemery commented on tests/helpers/test_condition.py:
```diff
@@ -2837,6 +2837,128 @@ async def test_if_action_before_and_after_during(hass, hass_ws_client, calls):
+    with patch("homeassistant.util.dt.utcnow", return_value=now):
+        hass.bus.async_fire("test_event")
+        await hass.async_block_till_done()
+        assert len(calls) == 3
```
```suggestion
        assert len(calls) == 4
```

janick replied:
Look good

emontnemery reviewed (commented):
With these changes, the test passes, but I think @janick should review it's as intended.

frenck commented:
@janick You have marked the above suggestions by emontnemery as resolved, but haven't committed any change. Did something go wrong?

../Frenck

janick commented:
@frenck Marked them unresolved. Is @emontnemery going to apply the changes or am I expected to replicate them in apply them myself??

frenck commented:
Some options:
- You can commit his suggestions one by one (with the commit button on each suggestion
- You can go to the files tab, and add each of the suggestions to the batch to commit. Next, scroll back up and commit the batch in a single commit.
- You can create a new commit manually in the IDE/editor you prefer and push the commit up.

It's up to you which you like.

../Frenck

janick committed:
Update tests/helpers/test_condition.py

Co-authored-by: Erik Montnemery <erik@montnemery.com>

janick committed:
Update tests/helpers/test_condition.py

Co-authored-by: Erik Montnemery <erik@montnemery.com>

janick committed:
Update tests/helpers/test_condition.py

Co-authored-by: Erik Montnemery <erik@montnemery.com>

janick committed:
Update tests/helpers/test_condition.py

Co-authored-by: Erik Montnemery <erik@montnemery.com>

frenck reviewed (approved):
Thanks, @janick 👍
